### PR TITLE
Improving performance of embulk-output-oracle

### PR DESCRIPTION
**Pull request by hito4t**
Support direct path insert via JDBC driver and via native client library.

### CODE REVIEW

**Comment by frsyuki** [comment]
* lib/linux_x64/libembulk-output-oracle.so is dynamically linked shared binary. Does it depend on certain Linux distribution (≒ glibc and libstdc++ version)?
* When I fix the code, how can I build the binary? Could you include a quick guide how to create the a environment?
* lib/* is global namespace of Ruby. Could you move embulk-output-oracle/lib/*/* binary files to embulk-output-oracle/lib/embulk/*?
* Generic function names such as `commit` and `rollback` are exported in the shared library. Does it cause function name conflicts in JVM process?
* It seems that the shared library is loaded into JVM globally. But the OracleOutputPlugin is loaded every time when a plugin is loaded using different ClassLoader. If a process runs OracleOutputPlugin twice using a single process, doesn't it cause problems?
* The biggest problem is that I can't fix the code when I get feedback or pull-requests. Why don't you join to this embulk-output-jdbc project as a committer?

If it uses jnr-ffi instead of JNI, packaging will be much easier. I think it's a TODO.

File: embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/OCI.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            } else if (osArch.equals("x86")) {
+                libFolder = new File(lib, "linux_x86");
+            }
+        }
+
+        if (libFolder != null) {
+            File libFile = new File(libFolder, libraryName);
+            if (libFile.exists()) {
+                logger.info(String.format("OCI : load library \"%s\".", libFile.getAbsolutePath()));
+                System.load(libFile.getAbsolutePath());
```
Comment (xerial) [inline embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/OCI.java]: Does it work when libFile is JarPath (e.g., embulk-plugin.jar!/lib/x86_64/.../xxx.so)?

**Comment by hito4t** [comment]
Thank you for review!

>    How much performance improves with the native shared library?
In my test environment, loading by the native libaray is about 3 times faster than JDBC driver.
(Total time is not so faster because reading and parsing will take time.)

>    lib/linux_x64/libembulk-output-oracle.so is dynamically linked shared binary. Does it depend on certain Linux distribution (≒ glibc and libstdc++ version)?
It depends on Oracle library as well as libstdc++, and the Oracle library depends on many other libraries.
The Oracle library may depend on cetain Linux distribution, so libembulk-output-oracle.so also does.

>    When I fix the code, how can I build the binary? Could you include a quick guide how to create the a environment?
OK. I'll create the guide.

>    lib/* is global namespace of Ruby. Could you move embulk-output-oracle/lib// binary files to embulk-output-oracle/lib/embulk/*?
I see. I'll move them.

>    Generic function names such as commit and rollback are exported in the shared library. Does it cause function name conflicts in JVM process?
There's no conflict now, but it may cause confilicts in future.
So I'll change the function names.

>    It seems that the shared library is loaded into JVM globally. But the OracleOutputPlugin is loaded every time when a plugin is loaded using different ClassLoader. If a process runs OracleOutputPlugin twice using a single process, doesn't it cause problems?
Yes, JVM cannot load same library by different ClassLoader and java.lang.UnsatisfiedLinkError will be thrown.
Does jnr-ffi solve this problem?

By the way, Oracle JDBC driver has two modes, Thin and OCI. Because OCI uses a native library, same problem will occur. For example, we cannot use Oracle JDBC OCI driver in both input plugin and output plugin.

>    The biggest problem is that I can't fix the code when I get feedback or pull-requests. Why don't you join to this embulk-output-jdbc project as a committer?
I see. I'm pleased to join!

> If it uses jnr-ffi instead of JNI, packaging will be much easier. I think it's a TODO.
I see. I'll add the issue.

**Comment by xerial** [comment]
> It seems that the shared library is loaded into JVM globally. But the OracleOutputPlugin is loaded every time when a plugin is loaded using different ClassLoader. If a process runs OracleOutputPlugin twice using a single process, doesn't it cause problems?
Yes, JVM cannot load same library by different ClassLoader and java.lang.UnsatisfiedLinkError will be thrown.

To resolve this problem, extract .so file as a temporary file and name it using a unique name for each class loader. Here is an example code used in snappy-java https://github.com/xerial/snappy-java/blob/develop/src/main/java/org/xerial/snappy/SnappyLoader.java#L205-L207

If different file name is used, we can load the same .so file multiple times with `System.load(...)`

File: embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/OCI.java (hunk 1 above)
Comment (hito4t) [inline embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/OCI.java]: Thank you for your comment!

The .so file is not in JAR file, it is in Ruby gem file and extracted when the plugin is installed.
The location of the .so file is as follows.
<pre>
embulk-output-oracle-X.X.X (plugin root)
└classpath
　├embulk-output-oracle-X.X.X.jar
　└lib
　　└…
　　　└libembulk-output-oracle.so
</pre>

First, OCI class gets its location (it works even if OCI class is in JAR file).
And finds the plugin root, then find .so file.

Comment (xerial) [inline embulk-output-oracle/src/main/java/org/embulk/output/oracle/oci/OCI.java]: That's good to know. Then there should be no problem. As I commented below, to support repeated load, we need to copy .so file to somewhere using a different name, then use `System.load(file)`

**Comment by hito4t** [comment]
Thank you!

It's a good solution, I'll try it.
In the case, it may be better to put the native library in JAR file than to put it out of the JAR file.

**Comment by hito4t** [comment]
> To resolve this problem, extract .so file as a temporary file ...

By the way, can we delete the temporary file?
I tried to delete by using File#deleteOnExit and File#delete in Runtime#addShutdownHook, but both didn't work.
I think we can't delete library files unloaded, but I can't find a way to release loaded libraries.

**Comment by xerial** [comment]
How about simply calling File.deleteOnExit immediately after creating the temporary file? I think no need exists to use shutdown hook because there is no guarantee that it is called properly (e.g., when JVM crashes)

**Comment by hito4t** [comment]
I thought so too, and I called File.deleteOnExit, but the file was not deleted...
JVM may delete files before releasing loaded library (at least in my environment).

I should think of another way.
For example, deleting old temporary files when plugin starts next time.

**Comment by hito4t** [comment]
I tried File.deleteOnExit on Linux Open JDK 7, then it works.
But I tried on Windows Oracle JDK 8, then it doesn't work.

**Comment by hito4t** [comment]
I've modified the sources and binaries.

* Build native library using old Oracle Client library (11gRelease1).
* Remove Japanese comments (auto generated by Visual Studio).
* Move native libraries to embulk-output-oracle/lib/embulk/*.
* Rename generic function names such as "commit" or "rollback".
* Copy native library before loading so that multiple plugins can run at the same time.
* Add issue about using jnr-ffi instead of JNI.

Would you check them?

I've not created the guide for creating a build environment yet.
I'll create it soon.

**Comment by frsyuki** [comment]
`File#deleteOnExit` causes memory leak. I highly recommend to not use it.

**Comment by frsyuki** [comment]
> Yes, JVM cannot load same library by different ClassLoader and java.lang.UnsatisfiedLinkError will be thrown.
> Does jnr-ffi solve this problem?
No, jnr-ffi doesn't solve. Application code needs to handle the problem because plugin load mechanism is Embulk's specific behavior.
A possible solution is to search C functions first. If the functions are not found, you can load the functions. It also need synchronization with a global object like `synchronize(System.class)` to be thread-safe. So, I still highly recommend you to use jnr-ffi and remove C code.

> Rename generic function names such as "commit" or "rollback".
I think general coding standard adds prefix to all non-static function names like `embulk_output_oracle_commit` or `embulk_output_oracle_free_handles`.

> I see. I'm pleased to join!
Thank you. I added you to embulk-output-jdbc committer group. Welcome to the embulk-output-jdbc group.

Above 2 concerns are still remaining. In our use case, because we use embulk as a module in a server program, it probably becomes a problem. But I think that it's mostly ok for command-line users.

Please merge the changes to the master branch. You can keep this ticket opened, or please create another pull-request when you work on above issues.
What's your email address on RubyGems? I'll add release permission of embulk-output gems to you.

**Comment by hito4t** [comment]
I've merged this pull-request.

> I still highly recommend you to use jnr-ffi and remove C code.

I see.
I've add the issue #20 .

> I think general coding standard adds prefix to all non-static function names like embulk_output_oracle_commit or embulk_output_oracle_free_handles.

I see.
I'll modify the function names soon.

> What's your email address on RubyGems? 

My email address is as follows.
thitoshi@cac.co.jp

**Comment by frsyuki** [comment]
Thank you. I added you to release managers of embulk-output-jdbc, mysql, postgresql, and redshift gems.

**Pull request merged**